src/material_library.h

```diff
@@ -0,0 +1,190 @@
+#include "material.h"
+#endif
+
+namespace pyne {
+
+typedef std::unordered_map<std::string, pyne::Material*> mat_map;
+typedef std::set<std::string> matname_set;
+typedef std::set<int> nuc_set;
+
+class MaterialLibrary {
+ protected:
+  // The actual library
+  mat_map material_library;  // material library
+  matname_set keylist;
+  nuc_set nuclist;
+
+
+ private:
+  /**
+   * \brief Add the nuclide form the material to the list of nuclides
+   * \param material from which add the nuclide to the list
+  */
+  void append_to_nuclist(pyne::Material mat);
+  /**
+   * \brief determines in that datapath exists in the hdf5 file
+   * \param[in] filename of the h5m file
+   * \param[in] the datapath of we would like to test
+   * \return true/false
+   */
+  bool hdf5_path_exists(const std::string& filename,
+                        const std::string& datapath);
+
+  /**
+   * \brief determines the length of an hdf5 data table
+   * \param[in] filename of the h5m file
+   * \param[in] the datapath of we would like to read
+   * \return the number of elements to the array
+   */
+  int get_length_of_table(const std::string& filename,
+                          const std::string& datapath);
+
+ public:
+  // materialLibrary constructor
+  MaterialLibrary();  //< empty constryctor
+
+  /**
+   * Constructor from file
+   * \param filename path to file on disk, this file may be either in plaintext
+   *                 or HDF5 format.
+  */
+  MaterialLibrary(const std::string& filename,
+                  const std::string& datapath = "/materials");
+
+  ~MaterialLibrary() {};  //< default destructor
+
+  /**
+   * \brief loads the pyne materials in map of name vs Material
+    /// \param filename Path on disk to the HDF5 file.
+    /// \param datapath Path to the materials in the file.
+    /// \param protocol Flag for layout of material on disk.
+  */
+  void from_hdf5(const std::string& filename,
+                 const std::string& datapath = "/materials",
+                 const std::string& nucpath = "", int protocol = 1);
+  /**
+   * \brief loads the pyne materials in map of name vs Material
+    /// \param filename Path on disk to the HDF5 file.
+    /// \param datapath Path to the materials in the file.
+    /// \param protocol Flag for layout of material on disk.
+  */
+  void from_hdf5(char* filename, char* datapath = "/materials" , char* nucpath = "", int protocol = 1);
+  /**
+   * \brief loads the pyne materials in map of name vs Material
+    /// \param filename Path on disk to the json file.
+  */
+  void from_json(char* filename);
+  /**
+   * \brief loads the pyne materials in map of name vs Material
+    /// \param filename Path on disk to the json file.
+  */
+  void from_json(const std::string& filename);
+  void load_json(Json::Value json);
+  Json::Value dump_json();
+  void write_json(const std::string& filename);
+  void write_json(char* filename);
+  /**
+   * Writes MaterialLibrary out to an HDF5 file.
```

Comment:
I think we need better documentation given some of the things we have discussed.
e.g. will this overwrite an existing `MaterialLibrary` in the file?

There are probably similar questions for some of the other functions, e.g. how does `merge` deal with conflicts?

This could either be documented for each function, or in longer documentation for the entire class.  

I am not sure what our practice is regarding documenting both the Python and C++ API's?